### ISSUE TITLE
Associated type of super trait not resolved in dyn syntax.

### ISSUE THREAD

**Issue by Nashenas88** [rust-lang/rust-analyzer#10018]
The following code compiles on the rust playground (https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5e6a4e9c3869301701a4984ed1433bbe)
```
use futures; // 0.3.16
use std::pin::Pin;
use futures::future::FusedFuture;
use futures::FutureExt;

fn main() {
    let _fut: Pin<Box<dyn FusedFuture<Output = ()> + Send>> = Box::pin(futures::future::ready(()).fuse());
}
```

However rust-analyzer marks `Output` as unresolved. In this case, `FusedFuture` has `Future` as a super trait, and requires the associated type `Output` from the `Future` trait to be specified in order to compile.

**Comment by ChayimFriedman2** [comment]
Fixed now, needs a test.

**Comment by fgamador** [comment]
@ChayimFriedman2 I'd like to look into writing that test. Can you point me to the relevant code, or give me guidance on how to find it myself? I don't see anything obvious in your commits around Sep 30. I'm guessing it's somewhere in rust-analyzer/crates/ide-diagnostics/src/handlers (and tests). Maybe unresolved_assoc_item.rs? Thanks.

**Comment by ChayimFriedman2** [comment]
The proper place is probably crates/hir-ty/src/tests/regression/new_solver.rs, note that you will have a bit of work minimizing the test case.

**Comment by fgamador** [comment]
Thanks. Could you also tell me the commit that fixed the issue? I'd like to try undoing it to see the test fail, plus I want to understand the code better.

**Comment by ChayimFriedman2** [comment]
I don't know what commit exactly it is, but it's probably some part of the migration to the new solver.

**Comment by fgamador** [comment]
Oh, I assumed you were the one who fixed it. Any idea how to find the relevant code?

**Comment by ChayimFriedman2** [comment]
I don't recommend you to try to search the commit that fixes it, it's pretty meaningless and will be hard work.

**Comment by ShoyuVanilla** [comment]
Writing a test code that passes with the current `master` branch and fails with the [last release](https://github.com/rust-lang/rust-analyzer/releases/tag/2025-08-25) before the next-gen trait solver migration would suffice. This has been fixed while the migration to the next-gen trait solver, not some small, targeted change for this issue, so you have to look into tens of thousands of lines of code to see what would have fixed this.

**Comment by ShoyuVanilla** [comment]
> I'm guessing it's somewhere in rust-analyzer/crates/ide-diagnostics/src/handlers (and tests). Maybe unresolved_assoc_item.rs?

And that code merely visualizes and reports the type inference errors while normalizing associated types actually caused in the `hir-ty` crate.

**Comment by ChayimFriedman2** [comment]
I think we should relax a bit on the requirement for a test for every closed issue. If it has easy reproduction, great, but otherwise we have a lot of fixed issues and not closing them waiting for a test seems not good to me.

**Comment by ShoyuVanilla** [comment]
Yeah, I think checking whether they are fixed in the latest version would suffice to close those issues, but I have a small concern closing them as writing regression tests for them are *good first contributions* for newcomers.

**Comment by fgamador** [comment]
This builds and passes, but it also passes if I delete the `type Assoc;` line, which should fail, so I don't think it's right.

```
#[test]
fn issue_10018() {
    check_no_mismatches(
        r#"
//- minicore: option
trait Parent {
    type Assoc;
}

trait Child: Parent {}

fn main() {
    let _: Option<Box<dyn Child<Assoc = ()>>> = None;
}
"#,
    );
}
```

**Comment by fgamador** [comment]
Also, maybe we should close this issue and open another one that says "Add test for ..."?
